HARK/ConsumptionSaving/PerfForesightNumbaModel.py

```diff
@@ -0,0 +1,379 @@
+    mNrmNow = np.append(mNrmNow, mNrmNow[-1] + 1.0)
+    cNrmNow = np.append(cNrmNow, cNrmNow[-1] + MPCmin)
+
+    # If the artificial borrowing constraint binds, combine the constrained and
+    # unconstrained consumption functions.
+    if BoroCnstArt > mNrmNow[0]:
+        # Find the highest index where constraint binds
+        cNrmCnst = mNrmNow - BoroCnstArt
+        CnstBinds = cNrmCnst < cNrmNow
+        idx = np.where(CnstBinds)[0][-1]
+
+        if idx < (mNrmNow.size - 1):
+            # If it is not the *very last* index, find the the critical level
+            # of mNrm where the artificial borrowing contraint begins to bind.
+            d0 = cNrmNow[idx] - cNrmCnst[idx]
+            d1 = cNrmCnst[idx + 1] - cNrmNow[idx + 1]
+            m0 = mNrmNow[idx]
+            m1 = mNrmNow[idx + 1]
+            alpha = d0 / (d0 + d1)
+            mCrit = m0 + alpha * (m1 - m0)
+
+            # Adjust the grids of mNrm and cNrm to account for the borrowing constraint.
+            cCrit = mCrit - BoroCnstArt
+            mNrmNow = np.concatenate(
+                (np.array([BoroCnstArt, mCrit]), mNrmNow[(idx + 1) :])
+            )
+            cNrmNow = np.concatenate((np.array([0.0, cCrit]), cNrmNow[(idx + 1) :]))
+
+        else:
+            # If it *is* the very last index, then there are only three points
+            # that characterize the consumption function: the artificial borrowing
+            # constraint, the constraint kink, and the extrapolation point.
+            mXtra = cNrmNow[-1] - cNrmCnst[-1] / (1.0 - MPCmin)
+            mCrit = mNrmNow[-1] + mXtra
+            cCrit = mCrit - BoroCnstArt
+            mNrmNow = np.array([BoroCnstArt, mCrit, mCrit + 1.0])
+            cNrmNow = np.array([0.0, cCrit, cCrit + MPCmin])
+
+    # If the mNrm and cNrm grids have become too large, throw out the last
+    # kink point, being sure to adjust the extrapolation.
+    if mNrmNow.size > MaxKinks:
+        mNrmNow = np.concatenate((mNrmNow[:-2], np.array([mNrmNow[-3] + 1.0])))
+        cNrmNow = np.concatenate((cNrmNow[:-2], np.array([cNrmNow[-3] + MPCmin])))
+
+    # Calculate the upper bound of the MPC as the slope of the bottom segment.
+    MPCmax = (cNrmNow[1] - cNrmNow[0]) / (mNrmNow[1] - mNrmNow[0])  # return
+
+    # Add two attributes to enable calculation of steady state market resources.
+    ExIncNext = 1.0  # Perfect foresight income of 1  # return
+    # Relabeling for compatibility with addSSmNrm
+    mNrmMinNow = mNrmNow[0]  # return
+
+    # See the PerfForesightConsumerType.ipynb documentation notebook for the derivations
+    vFuncNvrsSlope = MPCmin ** (-CRRA / (1.0 - CRRA))  # return
+
+    # Minimum market resources plus next income is okay starting guess
+    m_init_guess = mNrmMinNow + ExIncNext
+
+    try:
+        mNrmSS = newton_secant(
+            searchSSfunc,
+            m_init_guess,
+            args=(PermGroFac, Rfree, ExIncNext, mNrmNow, cNrmNow),
+        )[0]
+    except:
+        mNrmSS = None
```

Comment:
check if it possible just to use the converge information from `newton_secant` to rewrite the try, except code block.
Do check when newton_secant fails (so that we know the reason why there is this try, except code block)

Reply:
I'm actually a bit surprised that newton_secant does _not_ fail for the PF model, since the "secant" for a line is degenerate.  But it seems to have gotten the right answer.